src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -50,11 +51,18 @@ internal partial class TextDocumentState
             this.sourceTextOpt = sourceTextOpt;
             this.textAndVersionSource = textAndVersionSource;
 
+            Services = documentServiceProvider ?? DefaultDocumentServiceProvider.Instance;
+
             // for now, let it re-calculate if anything changed.
             // TODO: optimize this so that we only re-calcuate checksums that are actually changed
             _lazyChecksums = new AsyncLazy<DocumentStateChecksums>(ComputeChecksumsAsync, cacheResult: true);
         }
 
+        /// <summary>
+        /// A <see cref="IDocumentServiceProvider"/> associated with this document
+        /// </summary>
+        public IDocumentServiceProvider Services { get; }
```

Comment:
The name here is odd.  why 'Services', but it exposes a ServiceProvider?  in other places, where we have used 'Services' as the name of a property, the type name is something like SolutionServices, or HostWorkspaceServices.  The property name then follows the type name sensibly.  This feels weird here to me.

Reply:
naming I will leave it to design meeting. I don't care much. using Services since that's what is used in Workspace

Reply:
Sure!  Are you keeping track of the stuff that needs to go to design meeting?  Or does that need to be handled by some other means?  Thanks! :)